### PR TITLE
Abstorage

### PR DESCRIPTION
**Pull request by rohe**
Included the possibility to use a persistent storage.
One consequence of the work was that KeyJar was split into KeyJar and KeyIssuer.

### CODE REVIEW

File: src/cryptojwt/jwt.py [hunk 1]
```diff
@@ -107,7 +108,7 @@ def receiver_keys(self, recv, use):
         :param use: What the keys should be usable for
         :return: A list of keys.
         """
-        return self.key_jar.get(use, owner=recv)
+        return self.key_jar.get(use, issuer_id=recv)
```
Comment (jschlyter) [inline src/cryptojwt/jwt.py:111]: General comment: The rename of `issuer` to `issuer_id` will cause a lot of grief with existing code using named arguments (like my code). Please reconsider this name change.

If very important, we could add aliases and support both variants.

File: src/cryptojwt/key_bundle.py [hunk 2]
```diff
@@ -153,7 +153,7 @@ def ec_init(spec):
 class KeyBundle:
     """The Key Bundle"""
 
-    def __init__(self, keys=None, source="", cache_time=300, verify_ssl=True,
```
Comment (jschlyter) [inline src/cryptojwt/key_bundle.py:156]: If we drop `verify_ssl` here, should we also drop it from `KeyBundle`?

File: src/cryptojwt/serialize/__init__.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
+
+class SimpleList():
```
Comment (jschlyter) [inline src/cryptojwt/serialize/__init__.py]: Suggest we add a pure abstract base class (e.g., `SerializeBase`) and let SimpleList be a non-astract version of this. However, I cannot see that `SimpleList` is used anywhere in the code, is it? Leftovers?

File: src/cryptojwt/serialize/item.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
+import json
```
Comment (jschlyter) [inline src/cryptojwt/serialize/item.py:1]: Unused code as well?

**Review by jschlyter — CHANGES_REQUESTED** [review]
Generally good. Addressing backwards compatibility would help a lot.

File: src/cryptojwt/jwt.py (hunk 1 above)
Comment (rohe) [reply src/cryptojwt/jwt.py:111]: I've felt the grief :-( so I know what you're talking about. I still went ahead with the change.
Aliases is a good middle ground proposal.

Comment (rohe) [reply src/cryptojwt/jwt.py:111]: The only problem is of course that we'll never get rid of `owner` that way.

File: src/cryptojwt/key_bundle.py (hunk 2 above)
Comment (rohe) [reply src/cryptojwt/key_bundle.py:156]: Yes ! It should be dropped everywhere. Backward compatibility have made me keep it in `KeyBundle` so far.

File: src/cryptojwt/serialize/__init__.py (hunk 3 above)
Comment (rohe) [reply src/cryptojwt/serialize/__init__.py]: Leftovers.

File: src/cryptojwt/serialize/item.py (hunk 4 above)
Comment (rohe) [reply src/cryptojwt/serialize/item.py:1]: Not really. I have a problem here and that is that all the storage stuff is really defined in oidcmsg. I don't want cryptojwt be dependent on oidcmsg (wrong way around) so there are no tests for storage usage in cryptojwt as it stands right now. Have contemplated adding something simple (shelve based for instance). 
Anyway this module should probably be moved to oidcmsg.

File: src/cryptojwt/jwt.py (hunk 1 above)
Comment (rohe) [reply src/cryptojwt/jwt.py:111]: One problem I wanted to get away from was that the attribute name varied between methods/functions. Some times being `owner` some times `issuer` but always being the same thing.

File: src/cryptojwt/serialize/item.py (hunk 4 above)
Comment (rohe) [reply src/cryptojwt/serialize/item.py:1]: I kept it because it's cryptojwt specific.

**Comment by jschlyter** [comment]
`KeyJar.__getitem__ ()` previously defined to return all keybundles that belong to an entity. Now they return something different, breaking things. How can we address this? This was previously (in my case) to add keys to a KeyBundle in a KeyJar.

**Comment by rohe** [comment]
KeyJar.__getitem__(item) now returns the KeyIssuer instance that is bound to the issuer identifier `item`.

To get the same behaviour as before you now have to do:

keyjar[item].get_bundles()

key jar[item].all_keys() will return all keys irrespective of in which key bundle they reside.

**Comment by jschlyter** [comment]
Should `KeyJar[x]` not raise `KeyError` if no such KeyIssuer exists? Silently creating one seems strange.

**Comment by rohe** [comment]
Yes it should.

File: src/cryptojwt/key_jar.py [hunk 5]
```diff
@@ -325,119 +306,136 @@ def match_owner(self, url):
...
+        if issuer_id is None:
+            res = {}
+            for _, _issuer in self._issuers.items():
+                kbs = _issuer.find(source)
+                if kbs:
+                    res[_issuer.name] = kbs
+        else:
+            _issuer = self._get_issuer(issuer_id)
+            if _issuer is None:
+                return []
```
Comment (jschlyter) [inline src/cryptojwt/key_jar.py]: The old code return None if the issuer was not found, we should probably do the same here

Comment (rohe) [reply src/cryptojwt/key_jar.py]: Changed this. Are we done now or have I missed something ?

Comment (jschlyter) [reply src/cryptojwt/key_jar.py]: I think we're good.

**Review by jschlyter — APPROVED** [review]

**Pull request merged**